### PR TITLE
[DEMO ONLY]: Approach 2 for SMARTCAST_IMPOSSIBLE

### PR DESCRIPTION
**Pull request by tgeng**
This PR is still WIP and will likely undergo significant changes.

### CODE REVIEW

**Comment by tgeng** [comment]
Hi, @mglukhikh here is the PR with a prototype implementation of approach 2 in https://docs.google.com/document/d/1Nu64qcFRJrwV0WBKhz25ZIwGMktXEvvPLcJBth-AEw0/edit?ts=609ac077#

**Comment by demiurg906** [comment]
Hi @tgeng 

I take a quick look on this PR and part which relates to stability looks more or less OK, but last three commits about local vars capturing looks very scary and not fully correct. Consider such example:
```
fun take(s: String?): String = "" // (1)
fun take(s: String): Int = 1 // (2)

fun delayedRun(block: () -> Unit) {}

fun test() {
    var s: String? = null

    if (s != null) {
        delayedRun {
            take(s) // <---------------------------
        }
        s = null
    }
}
```

In FE 1.0 call `take(s)`  resolves to `(1)`, but after removal of `s = null` it resolves to `(2)`. In your branch it's a `(1)` in both scenarios.

I think we should discuss internally how local vars should be processed properly, so could you please postpone last three commits and focus on others in this PR?

File: compiler/fir/tree/src/org/jetbrains/kotlin/fir/expressions/impl/FirExpressionWithSmartcastImpl.kt [hunk 1]
```diff
@@ -19,11 +19,13 @@ import org.jetbrains.kotlin.fir.types.FirTypeRef
...
 import org.jetbrains.kotlin.fir.visitors.transformSingle
+import org.jetbrains.kotlin.types.SmartcastStability
 
 internal class FirExpressionWithSmartcastImpl(
     override var originalExpression: FirQualifiedAccessExpression,
-    override val typeRef: FirTypeRef,
-    override val typesFromSmartCast: Collection<ConeKotlinType>
+    override val smartcastType: FirTypeRef,
+    override val typesFromSmartCast: Collection<ConeKotlinType>,
+    override var smartcastStability: SmartcastStability?
```
Comment (demiurg906) [inline compiler/fir/tree/src/org/jetbrains/kotlin/fir/expressions/impl/FirExpressionWithSmartcastImpl.kt:28]: IMO this property should be `val` (and `originalExpression` should be `val` too actually).

Expressions with smartcasts (as any other expressions) should not be revisited after they were resolved, so having some mutable state in this class is strange

File: compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/53.fir.kt [hunk 2]
```diff
@@ -33,7 +33,7 @@ fun case_3(x: Int) = ""
 fun case_3(x: Int?) = 10
 fun case_3() {
     if (case_3_prop != null) {
-        val z = case_3(<!DEBUG_INFO_EXPRESSION_TYPE("kotlin.Int?")!>case_3_prop<!>)
+        val z = case_3(<!DEBUG_INFO_EXPRESSION_TYPE("kotlin.Int? & kotlin.Int?")!>case_3_prop<!>)
```
Comment (demiurg906) [inline compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/53.fir.kt:36]: Can you please change `DEBUG_INFO_EXPRESSION_TYPE` reporter so it won't render types from unstable smartcasts?

**Review by demiurg906 — CHANGES_REQUESTED** [review]

**Comment by tgeng** [comment]
Thanks Dmitriy! I don't intend to merge this PR. The case you illustrated is exactly the critical downside mentioned in the doc. I agree that this approach is not sufficient. If we ever need this, we should do pre-analysis combined with this. BTW I have another PR #4382 that reports SMARTCAST_IMPOSSIBLE without local var handling (that is, local var are treated as always stable).

File: compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/53.fir.kt (hunk 2 above)
Comment (jsjeon) [reply compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/53.fir.kt:36]: I think this is the place: https://github.com/JetBrains/kotlin/blob/master/compiler/tests-common-new/tests/org/jetbrains/kotlin/test/frontend/fir/handlers/FirDiagnosticsHandler.kt#L179 ?

Comment (tgeng) [reply compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/53.fir.kt:36]: Thanks! I have updated #4382.

**Pull request closed without merging**